# **Multiple Linear Regression**
# Medical Cost Prediction & Feature Impact Analysis

---

## **Introduction**

This project focuses on building a deep understanding of how machine learning models can be used to model real world problems, specifically in the healthcare domain. Using the Medical Cost Personal Dataset, the goal is not only to predict insurance charges but also to analyze how different features influence model behavior and performance.

Rather than treating this as a simple regression task, this project is approached as a full machine learning experiment, where different modeling choices, preprocessing techniques, and assumptions are carefully tested and evaluated.

---

## **Project Aim**

The aim of this project is to develop, analyze, and compare multiple regression models to accurately predict medical insurance costs while gaining a strong conceptual and practical understanding of machine learning techniques.

---

## **Objectives**

1. Build a solid understanding of regression algorithms through hands-on implementation
2. Perform in-depth exploratory data analysis (EDA) to uncover patterns and relationships
3. Analyze the effect of key features (especially BMI and smoking) on medical costs
4. Apply and compare different preprocessing techniques:
    - Encoding categorical variables
    - Feature scaling
    - Log transformation (for skewed data)
5. Train and evaluate multiple models, such as:
    - Linear Regression
    - Polynomial Regression
    - Support Vector Regression (SVR)
    - Decision Tree
    - Random Forest
6. Perform model comparison and critical analysis, not just reporting scores
7. Understand why a model performs well or fails, not just the final result

---
## **Dataset Description**

The dataset includes the following features:

- age → Age of the individual
- sex → Gender
- bmi → Body Mass Index
- children → Number of dependents
- smoker → Smoking status
- region → Residential area
- charges → Medical insurance cost (target variable)

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/d3lhomi10/medical-cost-personal-dataset)

---

## **Problem Type**

This is a Supervised Machine Learning – Regression Problem, where the goal is to predict a continuous target variable (charges).

---

Project Approach

This project is structured as a complete ML pipeline, including:

1. Data Exploration (EDA)
2. Data Preprocessing
3. Feature Engineering
4. Model Training
5. Model Evaluation
6. Model Comparison
7. Insights & Conclusions

The focus is not only on building a working model, but on understanding each step deeply and justifying every decision.




# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("insurance.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB
None

SHAPE OF DATA
Rows: 1338, Columns: 7

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 1

DATA TYPES
age           int64
sex          object
bmi         float64
children      int64
smoker       object
region       object
charges     float64
dtype: object

NUMERICAL SUMMARY
               age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397 

# **Understaning the outliers** 

### bmi: 9 outliers

In [4]:
# Calculate fences for year
Q1 = dataset['bmi'].quantile(0.25)
Q3 = dataset['bmi'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['bmi'] < lower_fence) | (dataset['bmi'] > upper_fence)]

# Display the rows
print(year_outliers)

      age     sex    bmi  children smoker     region      charges
116    58    male  49.06         0     no  southeast  11381.32540
286    46  female  48.07         2     no  northeast   9432.92530
401    47    male  47.52         1     no  southeast   8083.91980
543    54  female  47.41         0    yes  southeast  63770.42801
847    23    male  50.38         1     no  southeast   2438.05520
860    37  female  47.60         2    yes  southwest  46113.51100
1047   22    male  52.58         1    yes  southeast  44501.39820
1088   52    male  47.74         1     no  southeast   9748.91060
1317   18    male  53.13         0     no  southeast   1163.46270


In [5]:
print(dataset["bmi"])

0       27.900
1       33.770
2       33.000
3       22.705
4       28.880
         ...  
1333    30.970
1334    31.920
1335    36.850
1336    25.800
1337    29.070
Name: bmi, Length: 1338, dtype: float64


Why I Kept BMI Outliers

The extreme values in the bmi feature were retained because they represent real-world cases of severe obesity rather than data errors. Although these values were identified as statistical outliers, they are medically valid and important for building a model that generalizes well to real-life scenarios.

Removing these observations would reduce the model’s ability to learn from rare but meaningful patterns in the data.

---

How I Handled BMI

Instead of removing outliers, I applied feature scaling to the numerical features (including BMI) when required.

For models sensitive to feature magnitude (e.g., Linear Regression, Polynomial Regression, SVR), Standardization was applied to normalize the feature range.
For tree-based models (Decision Tree, Random Forest), no scaling was applied, as these models are not affected by feature scale.

This approach ensures that extreme BMI values are preserved while maintaining stable model performance.

# **Understaning the outliers** 

### charges: 139 outliers

In [6]:
# Calculate fences for year
Q1 = dataset['charges'].quantile(0.25)
Q3 = dataset['charges'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
year_outliers = dataset[(dataset['charges'] < lower_fence) | (dataset['charges'] > upper_fence)]

# Display the rows
print(year_outliers)

      age     sex     bmi  children smoker     region      charges
14     27    male  42.130         0    yes  southeast  39611.75770
19     30    male  35.300         0    yes  southwest  36837.46700
23     34  female  31.920         1    yes  northeast  37701.87680
29     31    male  36.300         2    yes  southwest  38711.00000
30     22    male  35.600         0    yes  southwest  35585.57600
...   ...     ...     ...       ...    ...        ...          ...
1300   45    male  30.360         0    yes  southeast  62592.87309
1301   62    male  30.875         3    yes  northwest  46718.16325
1303   43    male  27.800         0    yes  southwest  37829.72420
1313   19  female  34.700         2    yes  southwest  36397.57600
1323   42  female  40.370         2    yes  southeast  43896.37630

[139 rows x 7 columns]


In [7]:
print(dataset["charges"])

0       16884.92400
1        1725.55230
2        4449.46200
3       21984.47061
4        3866.85520
           ...     
1333    10600.54830
1334     2205.98080
1335     1629.83350
1336     2007.94500
1337    29141.36030
Name: charges, Length: 1338, dtype: float64


Why I Kept Charges Outliers

High values in the charges feature were identified as outliers using statistical methods. However, further analysis revealed that these values are strongly associated with the smoker feature, indicating that they represent meaningful real-world patterns rather than noise.

Removing these values would eliminate critical information about the impact of smoking on medical costs, leading to an unrealistic model.

---

How I Handled Charges:

Two approaches were used to handle the skewness in the target variable:

1. Without Log Transformation

    - The original charges values were used directly.
    - This allows evaluation of how models perform on raw data with naturally skewed distributions.


2. With Log Transformation

    A log transformation was applied to the target variable:

    `y_log = np.log(y)`

    This helps:
    - Reduce skewness

    - Stabilize variance

    - Improve performance of linear and distance-based models

After prediction, results were transformed back using:

`y_pred = np.exp(y_pred_log)`

This dual approach allows a fair comparison between models trained on raw and transformed targets.

# **Cleaning the data** 

### remove the duplication (1)

1338 - 1 = 1337 Rows

shape = 1337 * 7

In [8]:
# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset.drop_duplicates().copy()

# Final result
print(dataset_clean)

      age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1337 rows x 7 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [9]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

In [10]:
print(X)

[[19 'female' 27.9 0 'yes' 'southwest']
 [18 'male' 33.77 1 'no' 'southeast']
 [28 'male' 33.0 3 'no' 'southeast']
 ...
 [18 'female' 36.85 0 'no' 'southeast']
 [21 'female' 25.8 0 'no' 'southwest']
 [61 'female' 29.07 0 'yes' 'northwest']]


In [11]:
print(y)

[16884.924   1725.5523  4449.462  ...  1629.8335  2007.945  29141.3603]


### Log Transformation vs. Feature Scaling

In data preprocessing, **Log Transformation** fixes the **shape** of your distribution, while **Feature Scaling** fixes the **scale/magnitude** of your features.

---

#### 1. Log Transformation
**When to use:** Use this when your data is **right-skewed** (a long tail of large values) or has **heteroscedasticity** (variance increases with the mean).

* **Key Purpose:** It compresses the range of large values and expands the range of small values to make the distribution more **Gaussian (Normal)**.
* **Common Use Cases:**
    * Financial data (Income, House Prices).
    * Biological counts or population data.
    * When the relationship between variables is **multiplicative** (e.g., exponential growth).
* **Effect on Outliers:** It significantly reduces the relative distance of extreme outliers.



---

#### 2. Feature Scaling
**When to use:** Use this when features have **different units** (e.g., Age in years vs. Annual Income in EGP).

* **Key Purpose:** It ensures that no single feature dominates the model simply because it has a larger numerical range.
* **Techniques:**
    * **Standardization (Z-score):** Centers data around a mean of 0 with a standard deviation of 1.
    * **Normalization (Min-Max):** Rescales data to a fixed range, usually $[0, 1]$.
* **Required for:**
    * **Distance-based algorithms:** KNN, K-Means, SVM (where distances between points matter).
    * **Gradient-Descent algorithms:** Linear/Logistic Regression, Neural Networks (converges much faster).
    * **Dimensionality Reduction:** PCA (which looks for maximum variance).



---

#### Summary Table

| Feature | Log Transformation | Feature Scaling |
| :--- | :--- | :--- |
| **Primary Goal** | Change the **distribution shape**. | Change the **range/magnitude**. |
| **Main Problem** | Skewness / Outliers. | Unit mismatch / Slow convergence. |
| **Model Type** | Useful for almost all models if data is skewed. | Essential for Distance & Gradient models. |
| **Note** | Data must be positive ($x > 0$). | Can be applied to any numerical data. |

**Pro Tip:** For **Tree-based models** (Decision Trees, Random Forests), feature scaling is generally **not required** because the model splits based on rank, but Log Transformation can still help if the *target* variable is highly skewed.

In [12]:
y_log = np.log(y)
print(y_log)

"""""
You MUST convert predictions back:
```
y_pred_log = model.predict(X_test)
y_pred = np.exp(y_pred_log)
```
"""""

[ 9.73417643  7.45330245  8.40053847 ...  7.39623314  7.60486709
 10.27991376]


'""\nYou MUST convert predictions back:\n```\ny_pred_log = model.predict(X_test)\ny_pred = np.exp(y_pred_log)\n```\n'

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [13]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[1,4,5])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

In [14]:
print(X)

[[1.0 0.0 0.0 ... 19 27.9 0]
 [0.0 1.0 1.0 ... 18 33.77 1]
 [0.0 1.0 1.0 ... 28 33.0 3]
 ...
 [1.0 0.0 1.0 ... 18 36.85 0]
 [1.0 0.0 1.0 ... 21 25.8 0]
 [1.0 0.0 0.0 ... 61 29.07 0]]


# **Splitting the dataset into the Train set and Test set**

In [16]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y_log , test_size = 0.2, random_state = 0)

In [17]:
"""
# 'threshold=np.inf' tells numpy never to summarize the array
np.set_printoptions(threshold=np.inf, linewidth=1000)

run line code above if you want to show all input not sample
"""
print(X_test)

[[1.0 0.0 1.0 ... 18 39.82 0]
 [1.0 0.0 1.0 ... 47 29.37 1]
 [0.0 1.0 1.0 ... 49 31.35 1]
 ...
 [0.0 1.0 1.0 ... 18 26.18 2]
 [0.0 1.0 0.0 ... 54 25.1 3]
 [1.0 0.0 1.0 ... 62 30.495 2]]


In [18]:
print(y_test)

[ 7.3987629   9.0534165   9.13670885 10.39048166  9.17411742  7.89392623
  7.69538951  7.12412796  7.90500961  9.41212456  9.02167345  9.74351487
  9.54498634  9.01422863  8.60748151  7.79895595  8.56937154  8.15781434
  8.80094929  9.59510396  7.39153144  9.48984104  7.13592538  7.8797715
  7.4233489   8.44840194  8.22486705  9.35469211  8.23127532 10.53117162
  8.99462902 10.76770351  9.26644534  9.93451535  9.58684966 10.50726464
  9.02967716 10.84338854  9.039323    7.53928606 10.41857858  7.96106318
  8.42599666 10.76841119 10.49541794  9.00279756  9.89087755  8.94672242
  9.19827467  8.87812944 10.08444915  8.32785251 10.06765668 10.61967793
 10.09523691  8.02969623  8.06199617  9.08559426  9.12051971 10.72989783
  9.43151008  9.56685746  8.3531649  10.23786072  8.27984812  8.08284568
  7.45684558 10.56479197  8.06479687  9.00733227 10.6504394   8.45950826
  8.44964047  9.53877817  9.47364332  9.48926631  8.87446266  8.95304874
  7.53906429  8.56741242  9.36378044  8.07515394  8.

# **Building Multiple Linear Regression Model**

## Training the Multiple Linear Regression model on the Training set

In [19]:
from sklearn.linear_model import LinearRegression
"""
OOP
regressor is object
LinearRegression() is class
LinearRegression() no need to enter any parameters
"""
regressor = LinearRegression()
# train our model
regressor.fit(X_train, y_train)

LinearRegression()

# **Preticting the Test set results**

In [20]:
"""
Note:
y_test: contains the real profits
y_pred: contains the predicted profits
"""
y_pred_log = regressor.predict(X_test).reshape(-1,1)

"""
Note:
we did it only to compare betwen predict profits with real profits
to see how our model work (correct or wrong)
we also convert y_pred_log to original using np.exp 
"""
y_pred = np.exp(y_pred_log)
np.set_printoptions(precision=2,suppress=True)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), np.exp(y_test).reshape(len(y_test), 1)), axis = 1))

[[ 3051.26  1633.96]
 [ 7959.89  8547.69]
 [ 9581.99  9290.14]
 [12411.79 32548.34]
 [ 7722.55  9644.25]
 [ 3071.24  2680.95]
 [ 2992.    2198.19]
 [ 2369.55  1241.56]
 [ 3156.16  2710.83]
 [12429.93 12235.84]
 [ 7615.5   8280.62]
 [15137.02 17043.34]
 [13998.58 13974.46]
 [ 7992.21  8219.2 ]
 [ 5395.43  5472.45]
 [ 4310.57  2438.06]
 [ 5178.18  5267.82]
 [ 4427.33  3490.55]
 [ 6038.23  6640.54]
 [14588.42 14692.67]
 [ 2736.98  1622.19]
 [12710.29 13224.69]
 [ 2714.85  1256.3 ]
 [ 3531.94  2643.27]
 [ 3042.49  1674.63]
 [ 5784.64  4667.61]
 [ 3751.55  3732.63]
 [13616.84 11552.9 ]
 [ 3973.01  3756.62]
 [20168.17 37465.34]
 [ 8227.67  8059.68]
 [53388.85 47462.89]
 [ 8652.17 10577.09]
 [10162.23 20630.28]
 [16128.12 14571.89]
 [18304.96 36580.28]
 [ 8954.09  8347.16]
 [20871.21 51194.56]
 [ 7140.59  8428.07]
 [ 3102.53  1880.49]
 [15061.49 33475.82]
 [ 3586.91  2867.12]
 [ 5006.02  4564.19]
 [68030.11 47496.49]
 [13890.45 36149.48]
 [ 7970.47  8125.78]
 [10583.63 19749.38]
 [ 8033.39  7

# **Evaluating the Model Performance**

In [21]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred_log)

0.7716377049579233

# **Model Comparison Table**

| Experiment              | Log Transformation | Feature Scaling (X) | Feature Scaling (y) | R² Score (Test) |
| ----------------------- | ------------------ | ------------------- | ------------------- | --------------- |
| Baseline Model          | No                 | No                  | No                  | 0.7530          |
| Log Transformation Only | Yes                | No                  | No                  | 0.7716          |
| Log + Scaling (X only)  | Yes                | Yes                 | No                  | 0.7716          |
| Log + Scaling (X and y) | Yes                | Yes                 | Yes                 | 0.7716          |


---

# **Important Insights**


1. Log transformation significantly improved model performance, increasing the R² score from 0.7530 to 0.7716. This indicates that the target variable (charges) is right-skewed and benefits from a more normally distributed target.
2. Feature scaling had no effect on model performance. This is expected because Multiple Linear Regression is not sensitive to feature scale in terms of prediction accuracy.
3. After applying log transformation, no further improvement was observed. This suggests that preprocessing alone cannot overcome the inherent limitations of the model.
4. The main performance gain came from addressing data distribution, not from modifying feature magnitudes.

### Why Multiple Linear Regression is Limited in This Problem

Assumption of Linear Relationships
1. Multiple Linear Regression assumes a strictly linear relationship between input features and the target variable.

    However, in this dataset:

    - The effect of BMI on charges is not purely linear
    - The impact of smoking creates sharp, non-linear jumps in cost

These patterns cannot be fully captured by a linear model.

---


# **Conclusions**

1. The best configuration for Multiple Linear Regression is applying log transformation on the target variable
2. Feature scaling is not necessary and does not impact performance
3. The model provides a reasonable baseline but fails to capture complex patterns in the data


### This experiment demonstrates that:

1. Model choice is more critical than applying multiple preprocessing techniques
2. Linear models are useful for establishing a baseline and understanding relationships
3. However, they are insufficient for capturing real-world complexity

### To achieve better performance, it is necessary to move toward:

1. Polynomial Regression (to capture non-linearity)
2. Tree-based models such as Decision Trees and Random Forests
